src/Cryptol/Parser/ParserUtils.hs

```diff
@@ -295,21 +300,51 @@ mkEApp es@(eLast : _) = at (eFirst,eLast) $ foldl EApp f xs
      becomes
      [ f, x ` { a = 2 }, y ]
   -}
-  cvtTypeParams e [] = [e]
+  cvtTypeParams e [] = pure (e :| [])
   cvtTypeParams e (p : ps) =
-    case toTypeParam p of
-      Just fs -> cvtTypeParams (EAppT e fs) ps
-      Nothing -> e : cvtTypeParams p ps
-
-  toTypeParam e =
-    case dropLoc e of
-      ETypeVal t -> case dropLoc t of
-                      TTyApp fs -> Just (map mkTypeInst fs)
-                      _         -> Nothing
+    case toTypeParam p Nothing of
+      Nothing -> NE.cons e <$> cvtTypeParams p ps
+
+      Just (fs,ss,rng) ->
+        if checkAppExpr e then
+          let e'  = foldr (flip ESel) (EAppT e fs) ss
+              e'' = case rCombMaybe (getLoc e) rng of
+                      Just r -> ELocated e' r
+                      Nothing -> e'
+           in cvtTypeParams e'' ps
+        else
+          errorMessage (fromMaybe emptyRange (getLoc e))
+                  [ "Explicit type applications can only be applied to named values."
+                  , "Unexpected: " ++ show (pp e)
+                  ]
+
+  {- Check if the given expression is a legal target for explicit type application.
+     This is basically only variables, but we also allow the parenthesis and
+     the phantom "located" AST node.
+   -}
```

Comment:
Are there cases where allowing parens makes sense?

Reply:
Not really sure... I couldn't think of a good rationale either way.  We could certainly disallow it.  I don't think I've ever seen that pattern in code anywhere.